pkgs/sdk/server/src/Internal/FDv2DataSources/FDv2StreamingDataSource.cs

```diff
@@ -0,0 +1,396 @@
+            var putData = FDv2ChangeSetTranslator.TranslatePutData(changeset, _log);
+            if (_dataSourceUpdates is IDataSourceUpdatesHeaders dataSourceUpdatesHeaders)
+            {
+                return dataSourceUpdatesHeaders.InitWithHeaders(putData.Data, headers);
+            }
+
+            return _dataSourceUpdates.Init(putData.Data);
+        }
+
+        private bool HandlePartial(FDv2ChangeSet changeset)
+        {
+            // TODO: Check if the data source updates implementation supports the transactional store interface.
+            // If it does, then apply the data using that mechanism.
+            // Alternatively only support the transactional interface and report an error.
+            var patches = FDv2ChangeSetTranslator.TranslatePatchData(changeset, _log);
+            var success = true;
+            foreach (var patch in patches)
+            {
+                success = _dataSourceUpdates.Upsert(patch.Kind, patch.Key, patch.Item);
+                if (!success)
+                {
+                    break;
+                }
+            }
+
+            return success;
+        }
+
+        private void HandleJsonError(string message)
+        {
+            _log.Error("LaunchDarkly service request failed or received invalid data: {0}", message);
+
+            var errorInfo = new DataSourceStatus.ErrorInfo
+            {
+                Kind = DataSourceStatus.ErrorKind.InvalidData,
+                Message = message,
+                Time = DateTime.Now
+            };
+            _dataSourceUpdates.UpdateStatus(DataSourceState.Interrupted, errorInfo);
+
+            _es.Restart(false);
+        }
+
+        private void HandleStoreError(string message)
+        {
+            var errorInfo = new DataSourceStatus.ErrorInfo
+            {
+                Kind = DataSourceStatus.ErrorKind.StoreError,
+                Message = message,
+                Time = DateTime.Now
+            };
+            _dataSourceUpdates.UpdateStatus(DataSourceState.Interrupted, errorInfo);
+            if (_storeStatusMonitoringEnabled) return;
+
+            // If the value before the set was false, then we log the warning.
+            // We want to only log this once per outage.
+            if (!_lastStoreUpdateFailed.GetAndSet(true))
+            {
+                _log.Warn("Restarting stream to ensure that we have the latest data");
+            }
+
+            _es.Restart(false);
+        }
+
+        private void OnMessage(object sender, MessageReceivedEventArgs e)
+        {
+            var parsed = FDv2Event.TryDeserializeFromJsonString(
+                e.Message.Name,
+                e.Message.Data,
+                out var evt,
+                out var error);
+
+            if (!parsed)
+            {
+                HandleJsonError(error);
+                return;
+            }
+
+            IFDv2ProtocolAction action;
+            lock (_protocolLock)
```

Comment:
In practice operations will not be overlapped between OnMessage and OnOpen, but it could be a different thread servicing the operation. Using a lock here will handle the memory barrier needs.